vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java

```diff
@@ -159,49 +153,76 @@ public Set<T> getSelectedItems() {
                 .unmodifiableSet(new LinkedHashSet<>(selected.values()));
     }
 
+    /**
+     * Returns an unmodifiable view of the selected item ids.
+     *
+     * The returned Set may be a direct view of the internal data structures of this class.
+     * A defensive copy should be made by callers when iterating over this Set and modifying
+     * the selection during iteration to avoid ConcurrentModificationExceptions.
+     */
+    protected Set<Object> getSelectedItemIds() {
+        return Collections.unmodifiableSet(this.selected.keySet());
+    }
+
     @Override
     public Optional<T> getFirstSelectedItem() {
         return selected.values().stream().findFirst();
     }
 
     @Override
     public void select(T item) {
-        if (isSelected(item)) {
+        Object itemId = getItemId(item);
+        if (isSelectedItemId(item)) {
             return;
         }
-        Set<T> selected = new HashSet<>();
+        Map<Object, T> selected = new HashMap<>();
         if (item != null) {
-            selected.add(item);
+            selected.put(itemId, item);
         }
-        doUpdateSelection(selected, Collections.emptySet(), false);
+        doUpdateSelection(
+                selected,
+                Collections.emptyMap(),
+                false
+        );
     }
 
     @Override
     public void deselect(T item) {
-        if (!isSelected(item)) {
+        Object itemId = getItemId(item);
+        if (!isSelectedItemId(itemId)) {
```

Comment:
I think it is does not matter if it `isiSelected(item)` or `isSelectedItemId(itemId)`. Both should give back the same result I believe.

Reply:
Both do indeed give the same result, but this way we can prevent that the `getItemId` is called multiple times:
* once inside the `isSelected(item)` 
* another time in the `doUpdatedSelection`